python/test/unit/mesh/test_mesh.py

```diff
@@ -405,3 +408,108 @@ def test_unit_hex_mesh_assemble():
+    if d == 2:
+        mesh = create_unit_square(MPI.COMM_WORLD, n, n,
+                                  ghost_mode=ghost_mode)
+    else:
+        mesh = create_unit_cube(MPI.COMM_WORLD, n, n, n,
+                                ghost_mode=ghost_mode)
+
+    edim = mesh.topology.dim - codim
+    entities = locate_entities(mesh, edim, marker)
+    submesh, vertex_map, geom_map = create_submesh(mesh, edim, entities)
+    submesh_topology_test(mesh, submesh, vertex_map, edim, entities)
+    submesh_geometry_test(mesh, submesh, geom_map, edim, entities)
+
+
+@pytest.mark.parametrize("d", [2, 3])
+@pytest.mark.parametrize("n", [2, 6])
+@pytest.mark.parametrize("boundary", [boundary_0,
+                                      boundary_1,
+                                      boundary_2])
+@pytest.mark.parametrize("ghost_mode", [GhostMode.none,
+                                        GhostMode.shared_facet])
+def test_submesh_boundary(d, n, boundary, ghost_mode):
+    if d == 2:
+        mesh = create_unit_square(MPI.COMM_WORLD, n, n,
+                                  ghost_mode=ghost_mode)
+    else:
+        mesh = create_unit_cube(MPI.COMM_WORLD, n, n, n,
+                                ghost_mode=ghost_mode)
+    edim = mesh.topology.dim - 1
+    entities = locate_entities_boundary(mesh, edim, boundary)
+    submesh, vertex_map, geom_map = create_submesh(mesh, edim, entities)
+    submesh_topology_test(mesh, submesh, vertex_map, edim, entities)
+    submesh_geometry_test(mesh, submesh, geom_map, edim, entities)
+
+
+def submesh_topology_test(mesh, submesh, vertex_map, entity_dim, entities):
+    # Check that creating facets / creating connectivity doesn't cause
+    # a segmentation fault
+    mesh_tdim = mesh.topology.dim
+    if entity_dim == mesh_tdim:
+        submesh.topology.create_entities(mesh_tdim - 1)
+        submesh.topology.create_connectivity(mesh_tdim - 1, 0)
+
+    # Some processes might not own or ghost entities
+    if len(entities) > 0:
+        mesh.topology.create_connectivity(entity_dim, 0)
+        mesh_e_to_v = mesh.topology.connectivity(entity_dim, 0)
+        submesh.topology.create_connectivity(entity_dim, 0)
+        submesh_e_to_v = submesh.topology.connectivity(entity_dim, 0)
+        for submesh_entity in range(len(entities)):
+            submesh_entity_vertices = submesh_e_to_v.links(submesh_entity)
+            # The submesh is created such that entities is the map from the
+            # submesh entity to the mesh entity
+            mesh_entity = entities[submesh_entity]
+            mesh_entity_vertices = mesh_e_to_v.links(mesh_entity)
+            for i in range(len(submesh_entity_vertices)):
+                assert(vertex_map[submesh_entity_vertices[i]]
+                       == mesh_entity_vertices[i])
+    else:
+        assert(submesh.topology.index_map(entity_dim).size_local == 0)
+
+
+def submesh_geometry_test(mesh, submesh, geom_map, entity_dim, entities):
```

Comment:
Would be good to have some tests that use a submesh in onwards functionality, e.g. MeshTags, FunctionSpace etc.